ddl/foreign_key.go

```diff
@@ -94,54 +97,46 @@ func onDropForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ err
 		return ver, errors.Trace(err)
 	}
 
-	var (
-		fkName model.CIStr
-		found  bool
-		fkInfo model.FKInfo
-	)
+	var fkName model.CIStr
 	err = job.DecodeArgs(&fkName)
 	if err != nil {
 		job.State = model.JobStateCancelled
 		return ver, errors.Trace(err)
 	}
+	return dropForeignKey(d, t, job, tblInfo, fkName)
+}
 
+func dropForeignKey(d *ddlCtx, t *meta.Meta, job *model.Job, tblInfo *model.TableInfo, fkName model.CIStr) (ver int64, err error) {
+	var fkInfo *model.FKInfo
 	for _, fk := range tblInfo.ForeignKeys {
 		if fk.Name.L == fkName.L {
-			found = true
-			fkInfo = *fk
+			fkInfo = fk
+			break
 		}
 	}
-
-	if !found {
+	if fkInfo == nil {
 		job.State = model.JobStateCancelled
 		return ver, infoschema.ErrForeignKeyNotExists.GenWithStackByArgs(fkName)
 	}
-
 	nfks := tblInfo.ForeignKeys[:0]
 	for _, fk := range tblInfo.ForeignKeys {
 		if fk.Name.L != fkName.L {
 			nfks = append(nfks, fk)
 		}
 	}
 	tblInfo.ForeignKeys = nfks
-
-	originalState := fkInfo.State
-	switch fkInfo.State {
-	case model.StatePublic:
-		// We just support record the foreign key, so we just make it none.
-		// public -> none
-		fkInfo.State = model.StateNone
-		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != fkInfo.State)
-		if err != nil {
-			return ver, errors.Trace(err)
-		}
-		// Finish this job.
+	ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	// Finish this job.
+	if job.IsRollingback() {
+		job.FinishTableJob(model.JobStateRollbackDone, model.StateNone, ver, tblInfo)
+	} else {
 		job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
-		job.SchemaState = fkInfo.State
-		return ver, nil
-	default:
-		return ver, dbterror.ErrInvalidDDLState.GenWithStackByArgs("foreign key", fkInfo.State)
 	}
+	job.SchemaState = model.StateNone
```

Comment:
Why is it `StateNone`?

Reply:
For drop column/index/foreign_key job, the final `job.SchemaState` are both `StateNone`